packages/dds/tree/src/shared-tree/transactionTypes.ts

```diff
@@ -0,0 +1,84 @@
+ * All clients will validate the constraints of a transaction when it is sequenced, so all clients will agree on whether the transaction succeeds or not.
+ * @public
+ */
+export type TransactionConstraint = NodeInDocumentConstraint; // TODO: Add more constraint types here
+
+/**
+ * A transaction {@link TransactionConstraint | constraint} which requires that the given node exists in the tree.
+ * @remarks The node must be in the document (its {@link TreeStatus | status} must be {@link TreeStatus.InDocument | InDocument}) to qualify as "existing".
+ * @public
+ */
+export interface NodeInDocumentConstraint {
+	readonly type: "nodeInDocument";
+	readonly node: TreeNode;
+}
+
+/**
+ * The status of a transaction on the tree view.
+ * @alpha
+ */
+export interface TransactionResult<TResult> {
+	/**
+	 * The value returned by the inner `transaction` function or the special {@link RunTransaction.rollback | rollback value}
+	 * (`Tree.runTransaction.rollback`) which means that the transaction was aborted.
+	 */
+	readonly result: TResult | typeof rollback;
+
+	readonly undoPreconditions?: readonly TransactionConstraint[];
+}
+
+/**
+ * Parameters for running a transaction on the tree view that applies one or more edits to the tree as a single atomic unit.
+ * @alpha
+ */
+export interface RunTransactionParams<TResult> {
+	/**
+	 * The function to run as the body of the transaction.
+	 * @returns The result of the transaction. The user provided result (TResult) can either be returned directly or
+	 * as part of the `TransactionResult` object which can include other properties.
+	 * It could return nothing (TResult == void) to indicate a successful transaction.
+	 *
+	 * At any point during the transaction, the function may return the special {@link RunTransaction.rollback | rollback value}
+	 * (`Tree.runTransaction.rollback`) to abort the transaction and discard any changes it made so far.
+	 */
+	readonly transaction: () => TransactionResult<TResult> | TResult;
+	/**
+	 * An optional list of {@link TransactionConstraint | constraints} that are checked just before the transaction begins.
+	 * If any of the constraints are not met when `runTransaction` is called, it will throw an error.
+	 * If any of the constraints are not met after the transaction has been ordered by the service, it will be rolled back on
+	 * this client and ignored by all other clients.
+	 */
+	readonly preconditions?: readonly TransactionConstraint[];
+}
+
+/**
+ * The status of a transaction on the tree view.
+ * @alpha
+ */
+export interface RunTransactionResult<TResult> {
+	/**
+	 * The value returned by the inner `transaction` function or the special {@link RunTransaction.rollback | rollback value}
+	 * (`Tree.runTransaction.rollback`) which means that the transaction was aborted.
+	 */
+	readonly result: TResult | typeof rollback | undefined;
```

Comment:
What is the motivation for the ` | undefined` here?
I might be missing something, but I think having it here allows runTransaction to discard the TResult returned by the `transaction` delegate instead of returning it, which would be a bug.

Reply:
This has changed in the latest iteration.